File: mono/tests/tailcall-virt.cs
/*
Test tail.callvirt.

Author:
    Jay Krell (jaykrell@microsoft.com)

Copyright 2018 Microsoft
Licensed under the MIT license. See LICENSE file in the project root for full license information.

This IL will exit with 0 for success, 1 for failure.
It is based on nearby C# that is based on nearby F#.

This test covers tail.callvirt.
Incidentally it is a mutual recursion, not a self-recursion,
no generics, no pointers, small function signature, no value types, etc.

To properly test this:
 - compile with csc -optimize so ret is next to call
 - change call to tail. call
 - Or use the IL.

Actually only test the IL.
*/

using System;
using System.Runtime.CompilerServices;
using static System.Runtime.CompilerServices.MethodImplOptions;

unsafe class A
{
    static int check (long stack1, long stack2)
    {
        return (stack1 == stack2) ? 0 : 1;
    }

    [MethodImpl (NoInlining)]
    public virtual int f1(int counter, long initial_stack, long current_stack)
    {
        int local;
        if (counter > 0)
            return f2(counter - 1, initial_stack, (long)&local);
        return check((long)&local, current_stack);
    }

    [MethodImpl (NoInlining)]
    public virtual int f2(int counter, long initial_stack, long current_stack)
    {
        int local;
        if (counter > 0)
            return f1(counter - 1, initial_stack, (long)&local);
        return check((long)&local, current_stack);
    }

    [MethodImpl (NoInlining)]
    public static void Main()
    {
        int stack;
        Environment.Exit(new A().f1(100, (long)&stack, 0));
    }
}

File: mono/tests/tailcall-virt.fs
// Test tail.callvirt.
//
// Author:
//     Jay Krell (jaykrell@microsoft.com)
//
// Copyright 2018 Microsoft
// Licensed under the MIT license. See LICENSE file in the project root for full license information.
//
// This F# will stack overflow or exit with 1 or 2.
//
// This test covers tail.callvirt.
// Incidentally it is a mutual recursion, not a self-recursion,
// no generics, no pointers, small function signature, no value types, etc.
//
// Actually only test the nearby related IL, but this is what it is based on
// and this motivates that tail.callvirt is worth handling -- F# produces it.
//
// Be sure to compile with --optimize- else the tailcall might be optimized away.

type A() =
    member this.f2 (a) = if a > 0 then this.f1(a - 1)
                         else 1
    member this.f1 (a) = if a > 0 then this.f2(a - 1)
                         else 2

let a = A()
printfn "%d" (a.f1(1000 * 1000 * 10))
